File: C++/2-Medium/00138._随机链表的复制/00138._随机链表的复制.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;

    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == nullptr) return NULL;

        unordered_map<Node*, Node*> copies;
        Node* current = head;
        while (current){
            copies[current] = new Node(current->val);
            current = current->next;
        }

        current = head;
        while(current){
            copies[current]->next = copies[current->next];
            copies[current]->random = copies[current->random];
        }

        return copies[head];
    }
};

int main() {
    // Example usage
    Node* node1 = new Node(7);
    Node* node2 = new Node(13);
    Node* node3 = new Node(11);
    Node* node4 = new Node(10);
    Node* node5 = new Node(1);

    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;

    node2->random = node1;
    node3->random = node5;
    node4->random = node3;
    node5->random = node1;

    Solution solution;
    Node* copiedListHead = solution.copyRandomList(node1);

    // Printing the copied list to verify the solution
    Node* current = copiedListHead;
    while (current) {
        cout << "Value: " << current->val << ", Random Value: ";
        if (current->random) {
            cout << current->random->val;
        } else {
            cout << "null";
        }
        cout << endl;
        current = current->next;
    }

    // Cleaning up the allocated nodes (in a real scenario, consider smart pointers to manage memory automatically)
    delete node1;
    delete node2;
    delete node3;
    delete node4;
    delete node5;
    
    // Also, delete the copied nodes
    while (copiedListHead) {
        Node* temp = copiedListHead;
        copiedListHead = copiedListHead->next;
        delete temp;
    }

    return 0;
}

File: Python/2-Medium/00138._随机链表的复制/00138._随机链表的复制.py
from typing import Optional

class ListNode:
    def __init__(self, x: int, next: 'ListNode' = None, random: 'ListNode' = None):
        self.val = int(x)
        self.next = next
        self.random = random

def copyRandomList(head: 'Optional[ListNode]') -> 'Optional[ListNode]':
    if not head:
        return None
    
    # 创建旧节点到新节点的映射
    old_to_new = {}
    
    # 第一遍遍历：复制所有节点
    current = head
    while current:
        old_to_new[current] = ListNode(current.val)
        current = current.next
    
    # 第二遍遍历：复制 next 和 random 指针
    current = head
    while current:
        # 如果 next 不为空，则设置新节点的 next
        if current.next:
            old_to_new[current].next = old_to_new[current.next]
        # 如果 random 不为空，则设置新节点的 random
        if current.random:
            old_to_new[current].random = old_to_new[current.random]
        current = current.next
    
    # 返回新链表的头节点
    return old_to_new[head]

# 由于链表及其 random 指针的特殊性，无法直接在此环境中运行并验证代码的正确性。
# 此代码应在本地环境中，配合具体的链表输入进行测试。
